invenio_records_permissions/generators.py

```diff
@@ -6,227 +6,192 @@
     def __init__(self):
-        super(_Deny, self).__init__()
+        """Constructor."""
+        super(AnyUser, self).__init__()
 
-    def excludes(self):
+    def needs(self, **kwargs):
+        """Enabling low-level Needs."""
         return [any_user]
 
-    def query_filter(self):
-        return ~Q('match_all')  # Match None
-
+    def query_filter(self, **kwargs):
+        """Match all in search."""
+        return Q('match_all')
 
-Deny = _Deny()
 
-
-class _Admin(_NeedClass):
+class Deny(Generator):
+    """Denies ALL users (except super users)."""
 
     def __init__(self):
-        super(_Admin, self).__init__()
-
-    def needs(self):
-        return [ActionNeed('admin-access')]
+        """Constructor."""
+        super(Deny, self).__init__()
 
+    def excludes(self, **kwargs):
+        """Preventing low-level Needs."""
+        return [any_user]
 
-Admin = _Admin()
+    def query_filter(self, **kwargs):
+        """Match None in search."""
+        return ~Q('match_all')
 
 
-class _RecordNeedClass(_NeedClass):
+class Admin(Generator):
+    """Allows users with admin-access (different from superuser-access)."""
 
     def __init__(self):
-        super(_RecordNeedClass, self).__init__()
-
-    def needs(self, record):
-        pass
-
-    def excludes(self, record):
-        pass
+        """Constructor."""
+        super(Admin, self).__init__()
 
+    def needs(self, **kwargs):
+        """Enabling low-level Needs."""
+        return [ActionNeed('admin-access')]
 
-class _RecordOwners(_RecordNeedClass):
 
-    def __init__(self):
-        super(_RecordOwners, self).__init__()
+class RecordOwners(Generator):
+    """Allows record owners."""
 
-    def needs(self, record):
-        owner_needs = []
-        for owner in record.get('owners', []):
-            owner_needs.append(UserNeed(owner))
-        return owner_needs
+    def needs(self, record=None, **kwargs):
+        """Enabling low-level Needs."""
+        return [UserNeed(owner) for owner in record.get('owners', [])]
 
-    def query_filter(self):
+    def query_filter(self, record=None, **kwargs):
+        """Filters for current identity as owner."""
         provides = g.identity.provides
         for need in provides:
             if need.method == 'id':
                 return Q('term', owners=need.value)
-        return None
-
-
-RecordOwners = _RecordOwners()
-
-
-class _DepositOwners(_RecordNeedClass):
+        return []
 
-    def __init__(self):
-        super(_DepositOwners, self).__init__()
-
-    def needs(self, record):
-        deposit_owners = []
-        for owner in record.get('deposits', {}).get('owners', []):
-            deposit_owners.append(UserNeed(owner))
-        return deposit_owners
-
-    def query_filter(self):
-        provides = g.identity.provides
-        for need in provides:
-            if need.method == 'id':
-                return Q('term', **{"deposits.owners": need.value})
-        return None
 
+class AnyUserIfPublic(Generator):
```

Comment:
This looks good to me after the refactor. However, we should revisit when we deal with Files/Deposit because I believe that many parts are able to be generalized. Maybe add a #TODO